docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -0,0 +1,171 @@
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. They use the following
+address formats:
+
+- secp256k1: `ripemd160(sha256(pk_bytes))[:20]`
+- legacy amino multisig: `sha256(aminoCdc.Marshal(pk))[:20]`
+
+We don't want to change existing addresses. So the addresses for these two key types will remain the same.
+
+The current multisig public keys use amino serialization to generate the address. We will retain
+those public keys and their address formatting, and call them "legacy amino" multisig public keys
+in protobuf. We will also create multisig public keys without amino addresses to be described below.
+
+
+### Canonical Address Format
+
+We have three types of accounts we would like to create addresses for in the future:
+- regular public key addresses for new signature algorithms (ex. `sr25519`).
+- public key addresses for multisig public keys that don't use amino encoding
+- module accounts: basically any accounts which cannot sign transactions and
+which are managed internally by modules
+
+To address all of these use cases we propose the following basic `AddressHash` function,
+based on the discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694):
+
+```go
+func AddressHash(prefix string, contents []byte) []byte {
+	preImage := []byte(prefix)
+	if len(contents) != 0 {
+		preImage = append(preImage, 0)
+		preImage = append(preImage, contents...)
+	}
+	return blake2b.Sum256(preImage)[:20]
+}
+```
+
+`AddressHash` always take a string `prefix` as a starting point which should represent the
+type of public key (ex. `sr25519`) or module account being used (ex. `staking` or `group`).
+For public keys, the `contents` parameter is used to specify the binary contents of the public
+key. For module accounts, `contents` can be left empty (for modules which don't manage "sub-accounts"),
+or can be some module-specific content to specify different pools (ex. `bonded` or `not-bonded` for `staking`)
+or managed accounts (ex. different accounts managed by the `group` module).
+
+In the `preImage`, the byte value `0` is used as the separator between `prefix` and `contents`. This is a logical
+choice given that `0` is an invalid value for a string character and is commonly used as a null terminator.
+
+We use a 256-bit `blake2b` hash instead of `sha256` because it is generally considered more secure. Blake hashes
+are considered "random oracle indifferentiable", a stronger property which `sha256` does not have.
```

Comment:
@zmanian  I agree with your general approach. But we don't propose a any exotic algorithm. Blake2 is part of the at least Go and Python3 stdlib and it's an improvement. Some cryptographers claim that sha256 is easier to break than blake2 because we already know how to break other hash functions from the MD family.

Reply:
> Some cryptographers claim that sha256 is easier to break than blake2.

Some references would be appreciated :+1:

Reply:
> I'm just saying this because I'm always going to strongly advocate for a really high bar for adding a new hash function to the Cosmos trusted computing base.

So I want to summarize what I heard from @zmanian on our last call about this. Basically:
- there is no consensus on the replacement for SHA256 yet in the crypto community and there's a widespread sense that it won't be broken anytime soon
- we want to minimize what is needed for someone to implement "cosmos" on a broad array of devices including hardware wallets and embedded devices

Thus for now it's probably preferable to stick with SHA256. Is that accurate @zmanian ?

@alessio any progress on getting the cryptographer at AiB to take a look at this?

Reply:
W3F Research  / Polkadot is using blake2b and avoiding sha2.